py/persipubsub/queue.py

```diff
@@ -0,0 +1,489 @@
+                msgs_to_delete_timeout.add(key)
+
+    msgs_to_delete = msgs_to_delete.union(msgs_to_delete_timeout)
+    with queue.env.begin(write=True) as txn:
+        for key in msgs_to_delete:
+            txn.delete(key=key, db=pending_db)
+            txn.delete(key=key, db=meta_db)
+            txn.delete(key=key, db=data_db)
+
+    with queue.env.begin(write=True) as txn:
+        for sub_id in subscriber_ids:
+            sub_db = queue.env.open_db(
+                key=persipubsub.str_to_bytes(sub_id), txn=txn, create=False)
+            for key in msgs_to_delete_timeout:
+                txn.delete(key=key, db=sub_db)
+
+
+class _Queue:
+    """
+    Represent a message queue.
+
+    The messages come from many publishers and many subscribers.
+    They are stored persistently in an LMDB database.
+
+    :ivar path: to the queue
+    :vartype config_pth: pathlib.Path
+    :ivar env: LMDB environment on disk
+    :vartype env: lmdb.Environment
+    :ivar hwm: High water mark limit for queue
+    :vartype hwm: HighWaterMark
+    :ivar strategy:
+        Strategy which will be used to remove messages when high water mark is
+        reached.
+    :vartype strategy: Strategy
+    :ivar subscriber_ids: all subscribers of the queue
+    :vartype sub_list: List[str]
+    """
+
+    def __init__(self) -> None:
+        """Initialize class object."""
+        self.path = None  # type: Optional[pathlib.Path]
+        self.env = None  # type: Optional[lmdb.Environment]
+        self.hwm = None  # type: Optional[HighWaterMark]
+        self.strategy = None  # type: Optional[Strategy]
+        self.subscriber_ids = None  # type: Optional[List[str]]
+
+    def init(self,
+             path: Union[pathlib.Path, str],
+             env: Optional[lmdb.Environment] = None) -> None:
+        """
+        Initialize the queue.
+
+        :param path: where the queue is stored
+        :param env: open LMDB environment
+        """
+        # pylint: disable=too-many-arguments
+        # pylint: disable=too-many-locals
+        self.path = path if isinstance(path, pathlib.Path) \
+            else pathlib.Path(path)
+
+        if isinstance(env, lmdb.Environment):
+            self.env = env
+        else:
+            self.env = _initialize_environment(
+                queue_dir=self.path,
+                max_reader_num=persipubsub.MAX_READER_NUM,
+                max_db_num=persipubsub.MAX_DB_NUM,
+                max_db_size_bytes=persipubsub.MAX_DB_SIZE_BYTES)
+
+        with self.env.begin(write=True) as txn:
+            _ = self.env.open_db(key=persipubsub.DATA_DB, txn=txn, create=True)
+            _ = self.env.open_db(
+                key=persipubsub.PENDING_DB, txn=txn, create=True)
+            _ = self.env.open_db(key=persipubsub.META_DB, txn=txn, create=True)
+            _ = self.env.open_db(key=persipubsub.QUEUE_DB, txn=txn, create=True)
+
+        queue_data = persipubsub.lookup_queue_data(env=self.env)
```

Comment:
If this is the only place where persipubsub.lookup_queue_data is used, consider moving this function to this module.